Split layout for draftboards

natefaubion opened the pull request.

natefaubion commented:
@garyb I heard you liked monolithic PRs, so this one's for you.

The core data structure for the split layout is in `Draftboard.Pane`, and the layout concerns/invariants are in `Draftboard.Layout`. Otherwise `Draftboard.Component` has basically been rewritten, along with a lot of style changes to get the overlapping UIs to work. There are a lot of `pointer-event` hacks here to get all the interleaved elements working. This is obviously a major breaking change, so we should not merge this until we are ready to commit to backporting other bug fixes 😄.

natefaubion commented:
It's probably a good idea to actually try this PR out, to see if you have any concerns about the UI.

natefaubion commented on src/SlamData/Header/Gripper/Component.purs:
```diff
@@ -231,4 +234,8 @@ eval _ (Animated next) = do
+  H.get >>= raise ∘ H.action ∘ Notify
+  pure next
+
+eval _ (Notify _ next) =
```
I had to do this for the moment because this component is doing very bad things. This component should really all be in the `Header` component. After this is merged we should probably fix this.

natefaubion commented on src/SlamData/Header/Component.purs:
```diff
@@ -71,11 +71,15 @@ type DSL = H.ParentDSL State ChildState Query ChildQuery Slam ChildSlot
+    [ HP.classes
+        [ HH.className "sd-nav"
+        , HH.className if open then "open" else "closed" ]
+        ]
```
I had to add this status plus some other pointer-event hacks because inner decks need to be stacked on top of the gripper nub, but when you click on the gripper, it needs to be on top of the inner decks.

natefaubion commented on less/deck.less:
```diff
@@ -8,86 +8,114 @@
+      top: 0;
+      bottom: 0;
+      left: 0;
+      right: 0;
```
All of this contortion is to get the "floating" chrome on nested decks (with the root having normal chrome. Basically, In a deck's default state, the chrome is invisible, with the main UI taking up the full space. When you focus an inner deck, it all gets positioned _outside_ of its normal space using negative positioning, then the chrome is displayed with the main UI offset as usual. The transparent backgrounds are so the header nub can be seen, with the pointer event hacks ensuring it can still be clicked.

natefaubion commented on src/SlamData/Workspace/Card/Draftboard/Pane.purs:
```diff
@@ -0,0 +1,264 @@
+          Nil → Nil
+          all → List.singleton (one × Split orn all)
+      x : xs → go (i + 1) (x : pre) orn xs ix acc
+      _ → Nil
```
This might be dead code.

natefaubion commented on src/SlamData/Workspace/Card/Draftboard/Component.purs:
```diff
@@ -170,335 +99,330 @@ evalCard = case _ of
 evalBoard opts = case _ of
-  Grabbing deckId ev next → do
+  Init next → do
+    pure next
```
This is dead code

natefaubion commented on src/SlamData/Workspace/Card/Draftboard/Component/Render.purs:
```diff
@@ -0,0 +1,263 @@
+    ([ HP.classes
+         ([ HH.className "sd-draftboard-cell" ]
+          <> (HH.className "outer-edge-top" <$ guard (rect.top == 0.0))
+          <> (HH.className "outer-edge-left" <$ guard (rect.left == 0.0)))
```
I think these are unnecessary.

garyb commented:
> I heard you liked monolithic PRs, so this one's for you.

I probably deserved that :blush:. Since the treemap thing needs some discussion I'll get to reviewing now.

Regarding how/when we merge this, maybe we should start another branch, since the new chart types are still coming thick and fast - as we could still continue to release those as intermediate versions before this, as I assume this will break people's workbooks, not just that it's a big change to the UX? (and I don't think there will be conflicts between this and the charts).

garyb commented:
What are these comments about dead/unnecessary code by the way? If it's dead, is there a reason not to remove it?

natefaubion commented:
> What are these comments about dead/unnecessary code by the way? If it's dead, is there a reason not to remove it?

They were notes to myself :D I was reading over everything before signing off on Friday.

jdegoes commented:
>  I assume this will break people's workbooks, not just that it's a big change to the UX? (and I don't think there will be conflicts between this and the charts).

We should NOT break backward compatibility if possible. We can read the old or new format, and write the new format. If that's not possible with a reasonable amount of work, we can deal — but ideally every major version of the product would be stable (e.g. 3.x).

garyb commented:
I don't know whether what I said there is accurate, I was just making an assumption - @natefaubion?

natefaubion commented:
It's "possible", since both representations just use deck ids as pointers. Building an approximate layout has ambiguities with how you associate the splits, which may lead to some confusion. I can also just plop them in there for the user to sort out? Maybe its possible for me to get a hold of some of @damonLL's demo layouts.

natefaubion commented:
Alright, I've got a migration setup that works pretty well. I've tested it on some manufactured layouts and @damonLL's demo layouts that he sent me.

<img width="1082" alt="screen shot 2016-08-30 at 3 42 51 pm" src="https://cloud.githubusercontent.com/assets/144058/18106284/aeb78d42-6ec8-11e6-960b-09d86a5333b2.png">
<img width="711" alt="screen shot 2016-08-30 at 3 40 15 pm" src="https://cloud.githubusercontent.com/assets/144058/18106287/b49b5cac-6ec8-11e6-9c6d-a5743c15c209.png">
<img width="1060" alt="screen shot 2016-08-30 at 3 43 11 pm" src="https://cloud.githubusercontent.com/assets/144058/18106293/ba4f6b7a-6ec8-11e6-8268-b1940abc1ead.png">
<img width="709" alt="screen shot 2016-08-30 at 3 40 36 pm" src="https://cloud.githubusercontent.com/assets/144058/18106297/beb44186-6ec8-11e6-9dae-1f4014541aff.png">
<img width="1105" alt="screen shot 2016-08-30 at 3 43 38 pm" src="https://cloud.githubusercontent.com/assets/144058/18106300/c3cb5c54-6ec8-11e6-8761-41c94abb8fd3.png">
<img width="710" alt="screen shot 2016-08-30 at 3 40 51 pm" src="https://cloud.githubusercontent.com/assets/144058/18106307/c84ea6e6-6ec8-11e6-9fef-bb2363edb570.png">

natefaubion commented:
@garyb I think this is good to go!

natefaubion commented:
Addressing one point of feedback: Previously you could create a split at 5/9 and then 9/16, which results in a cell that often has no real dimension. It now validates that the split will create a cell at least 1/16 in size.

![disallow-small-splits](https://cloud.githubusercontent.com/assets/144058/18144312/9506d4d4-6f8b-11e6-9f0e-638457164462.gif)

natefaubion commented:
More tweaks from feedback.

Empty nested draftboards prompt you to zoom with the usual LOD indicator
<img width="786" alt="screen shot 2016-08-31 at 4 08 23 pm" src="https://cloud.githubusercontent.com/assets/144058/18146543/f3062a4e-6f95-11e6-8ded-b49aa79339e6.png">

Empty draftboards no longer have the `x` at the top right:
<img width="786" alt="screen shot 2016-08-31 at 4 08 36 pm" src="https://cloud.githubusercontent.com/assets/144058/18146554/ffbad276-6f95-11e6-8303-a3cb8ae8bd66.png">

Nested draftboards are treated as "published", such that the active cards are always set to the last available card, and all draftboard UI is hidden. Effectively "zoom to edit".
![nested-active-card](https://cloud.githubusercontent.com/assets/144058/18146571/15693c7a-6f96-11e6-8bb4-f27381cfd188.gif)

natefaubion commented:
I don't know what the deal is with the tests. Everything passes for me locally.

beckyconning commented:
It's weird that the screenshot isn't being taken

beckyconning commented:
Could the `throw` here be preventing the `logCurrentScreen`? https://github.com/slamdata/slamdata/blob/master/test/src/Test/Feature/Scenario.purs#L63

beckyconning commented:
I'm just thinking that having a screenshot would give us a better idea of what is going on.

garyb commented:
It's looking great with these revisions after our discussion yesterday!

garyb commented:
Ok, cool. I don't really have anything to comment on in the code, there's certainly nothing that jumped out at me as looking incorrect. :relaxed: 

It looks like there might be some minor styling issues with a couple of the cells now, but I don't know if we want to fix them now, or as separate PRs?

**Table cell**:

![table-cell-style](https://cloud.githubusercontent.com/assets/693642/18170814/996b48d0-7057-11e6-9d54-46623c7db704.png)

Looks like the overflowing element is in the wrong place now, it should just be the table container, but currently the table goes behind the pagination controls and the whole cell scrolls.

**Vars setup cell**:

![vars-style](https://cloud.githubusercontent.com/assets/693642/18170861/bfe636e6-7057-11e6-9ee6-f3104f9a9db7.png)

The lack of margin here looks pretty weird. I know @jdegoes said there should be no margin on the cards, to prevent excessive stacking, but this card probably shouldn't even be shown in published mode (not in the usual state anyway, maybe it should just list the vars instead), so maybe we can have them here?

natefaubion commented:
Some "content" like cards need to specify their own margin. I added some for markdown, FYI, because otherwise it was flush to the sides. I'm not sure what the deal is with jtable, I didn't _think_ I changed anything that would affect it, but who knows? Most of the core deck elements had `overflow: auto` on them, and I had to remove some of them for the floating chrome, but AFAIK I didn't touch anything specific to jtable (It still has overflow, just in the wrong place apparently?).

jdegoes commented:
We should fix these things now and try to make sure that, at least, published margins are visually pleasing and consistent.

@natefaubion How does this work interact with your previous "semantic grouping" project? Are we able to avoid stacking margins in nesting?

natefaubion commented:
@jdegoes margins are not stacked as there is no forced margin for decks/cards. Cards control what margin they have. When you nest a draftboard everything appears as though its on the same card (no separation betweek decks except whatever UI their cards provide).

natefaubion commented:
OK, cleaned up the variables card and the jtable overflow. I had to add an `!important` style, but we can clean this up when we do themes. There are some specificity issues.

<img width="997" alt="screen shot 2016-09-01 at 10 28 42 am" src="https://cloud.githubusercontent.com/assets/144058/18173373/f63fd162-702e-11e6-89a0-90afa1755733.png">
